libvast/src/meta_index.cpp

```diff
@@ -132,14 +132,22 @@ std::vector<uuid> meta_index::lookup(const expression& expr) const {
         return found_matching_synopsis ? result : all_partitions();
       };
       return caf::visit(detail::overload(
-        [&](const attribute_extractor& lhs, const data&) -> result_type {
+        [&](const attribute_extractor& lhs, const data& d) -> result_type {
           if (lhs.attr == system::time_atom::value) {
             auto pred = [](auto& field) {
               // FIXME: we should really just look at the &timestamp attribute
               // and not all fields of type time. [ch3843]
               return caf::holds_alternative<timestamp_type>(field.type);
             };
             return search(pred);
+          } else if (lhs.attr == system::type_atom::value) {
+            result_type result;
+            for (auto& [part_id, part_syn] : partition_synopses_)
+              for (auto& pair : part_syn)
+                if (evaluate(pair.first.name(), x.op, d))
+                  if (result.empty() || result.back() != part_id)
```

Comment:
There is no need to look at the rest of a partition once it was inserted in the result set. It would be better to skip the rest of the layouts. The `search` lambda has the same problem btw.

Reply:
Good point, I've reduced the number of loop iterations.